### DISCUSSION TITLE
PDF/UA experience?

### DISCUSSION THREAD

**Discussion by raphaelm** [py-pdf/pypdf#2694] [Q&A]
Is there anyone here with experience with PDF/UA? There seems to be almost none open tooling around it out there at all.

Are there freelancers among the amazing pypdf contributors that would be interested in investingating creating better open PDF/UA tooling with us (for money)?

**Comment by j-t-1** [comment]
> **AUGUST 12, 2024**: The PDF Association and leading PDF accessibility companies [Allyant](https://pdfa.org/member/allyant/), [Axes4](https://pdfa.org/member/axes4/), [CDP Communications](https://pdfa.org/member/cdp-communications/) and [TargetStream Technologies](https://pdfa.org/member/targetstream-technologies-inc/), are today making the ISO standards for accessibility in PDF technology - [ISO 14289-1 (PDF/UA-1)](https://pdfa.org/resource/iso-14289-pdfua/), [ISO 14289-2 (PDF/UA-2)](https://pdfa.org/iso-14289-2-pdfua-2/) and [ISO TS 32005](https://pdfa.org/resource/iso-32005/) - available at no cost.
> 
> Although users worldwide benefit from PDF’s accessibility features, adoption of accessibility standards for PDF technology has lagged because a significant portion of the ecosystem can’t or won’t pay for expensive ISO publications. Without a no-cost specification it’s very hard to reach all these developers, remediators and other users.
> 
> Starting today, all PDF users can begin to leverage ISO-standardized accessibility requirements for content in the PDF file format:
> 
> • **End users** who need assistive technology in order to navigate and read digital documents can use these standards to foster improved products and services.
> • **Organizations** can use these standards to set goals for the documents they create, ingest or manage.
> • **Document remediators** can use these standards in performing professional services related to ensuring PDF files are accessible.
> • **Software developers** can these standards as guidance in developing software to create and process tagged PDF.

https://pdfa.org/announcing-no-cost-access-to-pdfs-accessibility-standards/

**Comment by stefan6419846** [comment]
Where you able to find someone for this in the meantime and/or has there been any progress on the publicly available tooling in the meantime? At least as far as I know, not really?

**Reply by raphaelm** [reply]
I have not found something and am not aware of improvements.

**Comment by larsga** [comment]
I am interested in implementing this. At a minimum to get PyPDF to the point where it's possible to preserve tagged structure in input documents. Also want to implement support for copying pages from an input document and preserving the related tag structure.

If anyone has done work on this or has useful input I'd be happy to hear about it.

**Reply by stefan6419846** [reply]
```python3
reader = PdfReader('file.pdf')
writer = PdfWriter()
for page in reader.pages:
    writer.add_page(page)
writer.write('out.pdf')
```

already seems to preserve the tags as far as I can see. I have not tested more complex actions.

**Reply by larsga** [reply]
That's not what I observe with 5.4.0. 

I took a simple tagged test file, and ran it through your code above. The output is shown by Acrobat as being not tagged.

In the source file this `pdf.root_object['/MarkInfo']` returns an object containing `'/Marked': True`, showing the file is tagged. In the output, this is missing. This (`pdf.root_object['/StructTreeRoot']`) also does not return anything.

So as far as I can tell this is not correct. What are you observing that tells you the tags are preserved?

**Reply by stefan6419846** [reply]
I did not look at the root object or marker info, but at the tags inside the content stream itself. These are preserved as far as I can tell.

**Reply by larsga** [reply]
Aha, I see. So it may be enough to just write into the root object references to those tags in the pages, so that they can be found from there. If that's the case then maybe this isn't so much work, after all.

**Reply by larsga** [reply]
If I use `clone_from` in the `PdfWriter` then the output is marked as tagged, and the basic tagging structure is present. It still doesn't pass PDF/UA validation, but it gets a lot closer. I also don't know for sure what happens to pages merged in from other documents than the one that's cloned.

**Comment by j-t-1** [comment]
I am interested in helping, although cannot guarantee any time would be able to commit. If pypdf is used as a dependency, I could help with improvements of pypdf that would be useful downstream (and general enough to be appropriate for pypdf).

**Reply by larsga** [reply]
That's great to hear. The first step for my part will be a feasibility study to estimate how much work this would be -- basically to see if we can afford it or not. I'll update this thread when I know a little more, but it will be after Easter.

The idea is to contribute the improvement to PyPDF. As far as I can tell that's necessary to be able to write tags.

**Reply by stefan6419846** [reply]
As tags are more or less just plain PDF stream operations, this should work already. The only challenge might be to actually inject this into the streams afterwards.
